Document the agent and the device management API (#2235)
thin-edge/thin-edge.io: docs/src/references/agent/device-management-api.md

  - Initialized with all the required information to execute the command,
    they are updated during the command execution with outcome data.
  - A command request is cleared by publishing a retained empty string `""`.
- The command status messages contain a mandatory `"status"` field.
  - This status tells at which step the command execution is along its workflow. 
  - A new command is created with a `"status"` set to `"init"`.
  - When execution starts, the `"status"` is set to `"executing"`.
  - When execution is successful, the `"status"` is set to `"successful"`.
  - If execution fails for some reason, the `"status"` is set to `"failed"` and the reason given as `"reason"` field.
  - An operation might define specific `"status"` values for extra steps, actions or checks.

:::info
Health checks is a notable exception to the rule "command status are published as retained messages".
A health check is sent to the chanel `cmd/health/check` of a service (or a device) as a non-retained request `{}`,
triggering a health message response published on the `status/health` channel of the same service
(or respectively the channels of all the service running on that device).

> :::

Author: I have a concern about this use of the `cmd` channel for health check.

I would really prefer to send  health check requests on a `req` or `request` channel.
There are indeed too many differences:
- Health check requests must not be retained
- There is no need for a specific id for each request
- The response is sent to another topic (`status/health`).

Having such exceptions will make the implementation of extensible command workflow a bit more complex for no added value.

Having a `req/<request-type>` channel can be used for other kind of requests: [setpoints](https://en.wikipedia.org/wiki/Setpoint_(control_system)), clearing alarms ...

Reviewer: I wasn't a big fan of retrofitting health statuses by partially using the cmd APIs. But we did that at that point to avoid introducing yet another req/res mechanism just for health statuses. But, with additional things like setpoints or clear alarm requests, may be it's worth adding something.

But, I still wouldn't define an API that ties the response to the request in any way, as we want components to freely declare health status, setpoint values etc without any explicit request. Even for the clear alarm case, the prompt will be something and the resultant action will be just the regular "clear alarm" message. Will have to think about such a disjoint req/prompt mechanism.

Author: For details, see https://github.com/thin-edge/thin-edge.io/issues/2372

and, if so, to which extent.

```mermaid
graph LR
  te --/--- identifier --/--- cmd
  subgraph root
    te
  end

  subgraph target
    identifier["&lt;target identifier&gt;"]
  end

  subgraph command["command metadata"]
    direction LR
    cmd --/--- cmd_type["&lt;command_type&gt;"]
  end
```

Where the groups are described as follows:

| Group        | Description                                                                                                                                                                              |
|--------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| root         | The [topic root prefix](../mqtt-api.md/#group-root) for this installation of thin-edge.                                                                                                  |
| target       | The [topic identifier](../mqtt-api.md/#group-identifier) of the target device, possibly a [custom identifier](/contribute/design/mqtt-topic-design.md/#using-custom-identifier-schemas). |

> | `cmd`        | The [command channel](../mqtt-api.md/#channel-identifier) grouping of all the commands for this target.                                                                                  |

Reviewer 2: ```suggestion
| `cmd`        | The [command channel](../mqtt-api.md/#channel-identifier) grouping all of the commands for this target.                                                                                  |
```

A service that implements an operation for a device publishes on start a message notifying on the MQTT Bus
that this device can be sent commands of this type.

These messages are published with the retained flag set. So, a client process, such a mapper, can discover on start
what are __all the capabilities of all the devices__:

```sh te2mqtt formats="v1"
tedge mqtt sub 'te/+/+/+/+/cmd/+'
```

### Command status topics

The actual command requests are published on the __command status topics__.
For each request, a specific command topic is created to monitor the progress of the command from its initial state to its completion.
These topics are named using a unique command identifier forged by the requester.

```mermaid
graph LR
  te --/--- identifier --/--- cmd
  subgraph root
    te
  end

  subgraph target
    identifier["&lt;target identifier&gt;"]
  end

  subgraph command["command request"]
    direction LR
    cmd --/--- cmd_type["&lt;command_type&gt;"] --/--- cmd_id["&lt;cmd_id&gt;"]
  end
```

| Group        | Description                               |
|--------------|-------------------------------------------|

> | command_id   | The identifier of the operation instance. |

Reviewer 2: It might be worth mentioning that the `command_id` is an arbitrary string however it should be unique. It is recommended to either use a unique id generator, or add a unix timestamp as a suffix, e.g. `date +%s`

Author: Added

The messages published on these topics represent each the current status of a running command.

> So, one can list __all the pending commands of any type across all the devices__:

Reviewer 2: Though it is not just the pending operations but the in progress operations as well...so something like:

```suggestion
So, one can list __all the pending/in progress commands of any type across all the devices__:
```

  - Initialized with all the required information to execute the command,
    they are updated during the command execution with outcome data.
  - A command request is cleared by publishing a retained empty string `""`.
- The command status messages contain a mandatory `"status"` field.
  - This status tells at which step the command execution is along its workflow. 
  - A new command is created with a `"status"` set to `"init"`.
  - When execution starts, the `"status"` is set to `"executing"`.
  - When execution is successful, the `"status"` is set to `"successful"`.
  - If execution fails for some reason, the `"status"` is set to `"failed"` and the reason given as `"reason"` field.
  - An operation might define specific `"status"` values for extra steps, actions or checks.

:::info
Health checks is a notable exception to the rule "command status are published as retained messages".
A health check is sent to the channel `cmd/health/check` of a service (or a device) as a non-retained request `{}`,
triggering a health message response published on the `status/health` channel of the same service
(or respectively the channels of all the service running on that device).

> :::

Author: Health checks are implemented using a combination of 3 modes:
- passive mode: on start, the services publish an up status, on shutdown a down status
- crash mode: on crash, the MQTT broker publishes a down status on behalf of the dead process
- active mode: health check requests are sent on a regular pace (by a watchdog), and the services have to publish an up status.

## Operation workflow

A specific workflow rules each operation type, with specific:
- states
- message payloads
- states transitions.

However, there are some common rules:

> - There at least for states: `"init"`, `"executing"`, `"successful"` and `"failed"`.

Author: fixed

---
title: Device Management API
tags: [Reference, Agent, API]
sidebar_position: 2
---

# Device Management API

The agent and the operation plugins enable device management on thin-edge,
giving the cloud operators a uniform interface to install software packages,
update configuration files, inspect log files, restart devices ...

Each of these device management operation requires a different set of interactions between the participants,
with specific steps, message payloads and expectations.

Hence, each operation type comes with its own API, and, possibly, with different implementations,

> the `tedge-agent` service, operation plugins as the `tedge-log-plugin` or user-specific plugins.

Author: updated

  - Extra fragments can be added by participants and ignored by others if not relevant.
- Device capabilities are published as retained messages.
  - If a capability has no specific metadata, then the capability is published as an empty JSON object `{}`.
  - A capability is removed by publishing a retained empty string `""`.
- Command status messages are published as retained messages on their dedicated topic.
  - These messages represent the current state of the command.
  - Initialized with all the required information to execute the command,
    they are updated during the command execution with outcome data.
  - A command request is cleared by publishing a retained empty string `""`.
- The command status messages contain a mandatory `"status"` field.
  - This status tells at which step the command execution is along its workflow. 
  - A new command is created with a `"status"` set to `"init"`.
  - When execution starts, the `"status"` is set to `"executing"`.
  - When execution is successful, the `"status"` is set to `"successful"`.
  - If execution fails for some reason, the `"status"` is set to `"failed"` and the reason given as `"reason"` field.

>   - An operation might define specific `"status"` values for extra steps, actions or checks.

Reviewer: ```suggestion
  - Different operations might define other specific `"status"` values for extra steps, actions or checks.
```

Author: Updated along an intermediate suggestion:

```suggestion
  - Each operation might define other specific `"status"` values for extra steps, actions or checks.
```

- There only two terminal states: `"successful"` and `"failed"`.
- The retained messages are finally cleared by the requester.

```mermaid
flowchart LR
    init --> executing --> successful
    executing --> failed
    successful --> done
    failed --> done

    subgraph init_state["status: init"]
      init(gives all the data to proceed<br/><br/>created by the requester<br/>typically a mapper)
    end

    subgraph executing_state["status: executing"]
      executing(the command is under execution<br/>the content is the same as for init)
    end

    subgraph successful_state["status: successful"]
      successful(the command has been successful<br/>some outcome might be returned to the requester)
    end

    subgraph failed_state["status: failed"]
      failed(the command has failed<br/>the reason is given to the requester)
    end

    subgraph done_state[done]
      done(finally cleared by the requester)
    end
```

### Capability declaration

On start, a service that implements an operation for a device must:

- know the MQTT root prefix (per default: `te`),
- know the entity topic identifier of the target device (e.g. `device/child001//`),

> - know the well-known name of the operation (e.g. `software-update`),

Author: Fixed all over the new pages.

- forge a capability message describing the operation support (e.g. `{ "supportedTypes": ["apt", "docker"] }`)
- publish this capability message, with a retained flag, on the target command metadata topic (e.g. `te/device/child001///cmd/software-update`)

### Command execution

To request the execution of a command, a client must:
- know the MQTT root prefix (per default: `te`),
- know the entity topic identifier of the target device (e.g. `device/child001//`),
- know the well-known name of the operation (e.g. `software-update`),
- assign a unique id to its command (say `c8y-1234`)
- use these target and command identifiers to forge the topic name for its request

>   (e.g. `te/device/child001///cmd/software-update/c8y-1234`)

Reviewer: I'd just compress all that into a single statement: "subscribe to the command topic that includes the command id as well (e.g. `te/device/child001///cmd/software-update/c8y-1234`).

Using a term like command topic should be sufficient as we've already established its format in the previous section.

Author: Agree that this is bit verbose. But I think it helps to be explicit in a reference documentation on what needs to given to and then done by the operation plugins.

- describe the request in a JSON message as defined by the operation API
- publish, as a retained message, the request on its specific topic with a `"status"` set to `"init"`
- await on the same topic for a response message with a `"status"` set to `"executing"`, `"successful"` or `"failed"`
- ignore any message with an unknown status as the agent is free to use intermediate states
- finally, clear the command topic by sending an empty retained message

To implement an operation on behalf of a device, a service must:
- know the MQTT root prefix (per default: `te`),
- know the entity topic identifier of the target device (e.g. `device/child001//`),
- know the well-known name of the operation (e.g. `software-update`),
- subscribe to any command requests of this type and the target device (i.e. `te/device/child001///cmd/software-update/+`)

> - react to a well-formed requests by publishing the updated status when the step is done.

Reviewer: Again, not worth repeating I feel.

Author: Indeed, I would not repeat that all and every operation. However, this must be said once.

### Example

As an example, let's take software updates on a child device.

When launched on the child device `device/child001//`,

> `tedge-agent` notifies that software packages can be updated on this device:

Reviewer: ```suggestion
`tedge-agent` notifies that software packages of types: `apt` and `docker` can be updated on this device:
```
Just to highlight the main points in the example.

Author: Added

                    "name": "nodered",
                    "version": "1.0.0",
                    "action": "install"
                }
            ]
        }
    ]
}'
```

The agent, running on `device/child001//`, notifies that it will execute the command:

```sh te2mqtt formats="v1"
tedge mqtt pub --retain 'te/device/child001///cmd/software_update/c8y-123' '{
    "status": "executing",
    "updateList": [
        {
            "type": "debian",
            "modules": [
                {
                    "name": "nodered",
                    "version": "1.0.0",
                    "action": "install"
                }
            ]
        }
    ]
}'
```

Then the agent proceeds, here installing a specific version of `nodered`, and notifies the mapper when done:

```sh te2mqtt formats="v1"
tedge mqtt pub --retain 'te/device/child001///cmd/software_update/c8y-123' '{
    "status": "successful",
    "updateList": [
        {
            "type": "debian",
            "modules": [
                {
                    "name": "nodered",
                    "version": "1.0.0",
                    "action": "install"
                }
            ]
        }
    ],

>     "currentSoftwareList": [

Author: Oops!

>     "currentSoftwareList": [

Author: Fixed

        {
            "type": "debian",
            "modules": [
                {
                    "name": "nodered",
                    "version": "1.0.0",
                },
                {
                    "name": "collectd",
                    "version": "5.7"
                }
            ]
        },
        {
            "type": "docker",
            "modules": [
                {
                    "name": "nginx",
                    "version": "1.21.0",
                },
                {
                    "name": "mongodb",
                    "version": "4.4.6",
                }
            ]
        }
    ]
}'
```

The Cumulocity mapper, having subscribed to all software-update commands,
monitors this command instance and notifies Cumulocity of its progress upto completion.  
Finally, the Cumulocity mapper clear the command topic:

```sh te2mqtt formats="v1"
tedge mqtt pub --retain 'te/device/child001///cmd/software-update/c8y-123' ''

> ```

Reviewer: I'm assuming that the workflow API bits like the workflow definition file from [this PR](https://github.com/thin-edge/thin-edge.io/pull/2071) will also be added to the device-mangement-api doc at some point. Because, that'll give clarity to the users on how different components can work together with `tedge-agent` to drive an operation to completion.

Author: Yes, definitely. I even think the whole [operation workflow specification](https://github.com/thin-edge/thin-edge.io/pull/2071) has to reworked on top of this PR.
